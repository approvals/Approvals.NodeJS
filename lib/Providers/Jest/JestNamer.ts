import { expect } from "@jest/globals";
import path from "path";
const Namer:any = require( "../../Namer");

export function convertToFilename(name: string):string {
    const forbidden = "\\/:?\"<>|' ";
    for (let forbiddenKey of forbidden) {
        if (name.includes(forbiddenKey)) {
            name = name.replace(forbiddenKey, "_");
        }
    }
    return name;
}

export function getJestNamer(): typeof Namer{
    const state = expect.getState();
    const file = path.parse(state.testPath as string);
    const testPath = file.dir;
    const testFileName = file.name;
    const testName = convertToFilename(`${testFileName}.${state.currentTestName}`);
    return new Namer(testPath, testName);
}
