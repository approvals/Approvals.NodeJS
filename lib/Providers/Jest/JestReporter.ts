import {expect} from "@jest/globals";
import type {ApprovalFailureReporter} from "../../Core/ApprovalFailureReporter";

const fs = require('fs');

export class JestReporter implements ApprovalFailureReporter {
    name: string = "JestReporter";

    canReportOn() {
        return true;
    }

    report(approvedFilePath: string, receivedFilePath: string) {
        const approvedText = fs.readFileSync(approvedFilePath).toString();
        const receivedText = fs.readFileSync(receivedFilePath).toString();
        expect(receivedText).toBe(approvedText);
    }
}
