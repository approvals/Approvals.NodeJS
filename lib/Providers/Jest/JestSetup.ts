import { ApprovedFileLog } from "../../Logs/ApprovedFileLog";
import { FailedFileLog } from "../../Logs/FailedFileLog";
import fs from "fs";
import { LogUtils } from "../../Logs/LogUtils";

export let jestInitializingRun = false;

export function initializeGlobalsForJest() {
  ApprovedFileLog.forceClearLogFile();
  FailedFileLog.forceClearFileLog();
  jestInitializingRun = true;
}

export function helpUserSetupJest(fileToCheck: string) {
  LogUtils.ensureTempDirectoryExists();
  // Check if the file exists
  if (fs.existsSync(fileToCheck)) {
    // Get the date that fileToCheck was created
    const fileStats = fs.statSync(fileToCheck);
    const fileCreationDate = fileStats.birthtime;

    // Check if the date was in the last 24 hours
    const currentDate = new Date();
    const twentyFourHoursAgo = new Date(
      currentDate.getTime() - 24 * 60 * 60 * 1000,
    );
    const isFileRecent = fileCreationDate > twentyFourHoursAgo;

    if (isFileRecent) {
      return;
    }
  }
  const helpMessage = `
**********************************************
Jest needs to be set up for logging.
More information at https://github.com/approvals/Approvals.NodeJS/blob/master/docs/how_tos/InitializeJestForLogging.md
**********************************************
`;
  console.log(helpMessage);
}
