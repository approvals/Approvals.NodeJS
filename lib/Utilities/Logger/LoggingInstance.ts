import {StringWrapper} from "./StringWrapper";


class Toggles {
    public queries: boolean;
    public messages: boolean;
    public variables: boolean;
    public hour_glass: boolean;
    public events: boolean;
    public markers: boolean;

    constructor(show: boolean) {
        this.queries = show;
        this.messages = show;
        this.variables = show;
        this.hour_glass = show;
        this.markers = show;
        this.events = show;
    }
}

function printType(value:any) {
    return `<${value.constructor.name}>`;
}

function getCallingMethod(additional_stack: number) {
    const re = /at ([^(]+) \(/g;
    const stack = `${new Error().stack}`;
    const lines = stack.split("\n");
    const stackDepth = 2 + additional_stack;
    const line = lines[stackDepth]
    const aRegexResult = re.exec(line) ?? [];
    const name = aRegexResult[1] || aRegexResult[2];
    return name;
}

export class LoggingInstance {
    logger: (text: string) => void;
    private counter: number = 0;
    private tabs: number = 0;
    private toggles = new Toggles(true);
    public log_stack_traces: boolean = true;
    private log_with_timestamps: boolean = true;
    public timer: () => Date;
    private previous_timestamp: Date | null = null;

    constructor() {
        this.logger = console.log;
        this.timer = () => new Date();
    }

    log_to_string(): StringWrapper {

        const stringWrapper = new StringWrapper();
        this.log_with_timestamps = false;
        this.log_stack_traces = false;
        this.logger = (t) => stringWrapper.append(t)
        return stringWrapper;
    }

    use_markers(additional_stack: number, code: () => void, parameters: string | (() => string) ="" ) {
        if (!this.toggles.markers) {
            code();
            return;
        }
        const name = getCallingMethod(additional_stack + 1);
        let parameterText = "";
        if (typeof parameters === 'function'){
            parameterText = parameters();
        }
        else {
            parameterText = parameters;
        }
        this.log_line(`=> ${name}(${parameterText})`)
        this.withTabbing(code)
        if (typeof parameters === 'function'){
            parameterText = parameters();
        }
        else {
            parameterText = ""
        }
        this.log_line(`<= ${name}(${parameterText})`)
    }

    variable(name: string, value: any, showTypes: boolean) {
        if (!this.toggles.variables) {
            return;
        }

        let toType = (v: any, s = "") => ''
        if (showTypes) {
            toType = (value, spacing = " ") => `${spacing}${printType(value)}`
        }

        if (Array.isArray(value)) {
            this.log_line(`variable: ${name}${toType(value, '')}.length = ${value.length}`)
            this.withTabbing(() => {
                value.forEach((v, i) => {
                    this.logger(`${this.getTabs()}${name}[${i}] = ${v}${toType(v)}\n`);
                });
            });
        } else {
            this.log_line(`variable: ${name} = ${value}${toType(value)}`)

        }
    }

    private log_line(text: string, use_timestamps: boolean = true) {
        if (this.counter != 0) {
            this.logger("\n")
        }
        this.counter = 0
        const timestamp = use_timestamps ? this.get_timestamp() : "";
        const output_message = `${timestamp}${this.getTabs()}${text}\n`
        this.logger(output_message)

    }

    private getTabs() {
        return "  ".repeat(this.tabs);
    }

    private withTabbing(code: () => void) {

        this.tabs += 1;
        code();
        this.tabs -= 1;
    }

    hour_glass() {
        if (!this.toggles.hour_glass) {
            return;
        }

        this.counter += 1;
        if (this.counter == 1) {
            this.logger(`${this.getTabs()}.`);
        } else if (this.counter == 100) {
            this.logger("10\n")
            this.counter = 0
        } else if (this.counter % 10 == 0) {
            const digit = (this.counter / 10)
            this.logger(`${digit}`
            )
        } else {

            this.logger(".")
        }
    }


    show_all(show: boolean) {
        this.toggles = new Toggles(show);
    }

    event(event_name: string) {
        if (!this.toggles.events) {
            return;
        }
        this.log_line(`event: ${event_name}`)
    }

    show_queries(show: boolean) {
        this.toggles.queries = show
    }

    show_markers(show: boolean) {
        this.toggles.markers = show
    }

    show_events(show: boolean) {
        this.toggles.events = show
    }

    show_messages(show: boolean) {
        this.toggles.messages = show
    }

    show_variables(show: boolean) {
        this.toggles.variables = show
    }

    show_hour_glass(show: boolean) {
        this.toggles.hour_glass = show
    }

    warning(exception: Error | string) {
        const warning_stars = "*".repeat(91);
        const text = null;
        this.log_line(warning_stars, false);
        if (this.log_with_timestamps) {
            this.log_line("", true)
        }
        if (text) {
            this.log_line(`Message:${text}`, false)
        }
        if (exception) {
            let stack_trace = "";
            if (this.log_stack_traces) {
                // todo: grab stack trace
                stack_trace = exception.toString()

            } else {
                stack_trace = `${exception}`;
            }
            this.log_line(stack_trace,  false)
        }
        this.log_line(warning_stars,  false)
    }

    query(queryText: string) {
        if (!this.toggles.queries) {
            return;
        }
        this.log_line(`Sql: ${queryText}`)
    }

    message(messageText: string) {
        if (!this.toggles.messages) {
            return;
        }
        this.log_line(`message: ${messageText}`)
    }

    show_timestamps(show: boolean) {
        this.log_with_timestamps = show;

    }

    private get_timestamp() {
        if (! this.log_with_timestamps){
            return "";
        }

        const time1 = this.timer();
        const time = time1.toISOString();
        let diff_millseconds = 0;
        if (this.previous_timestamp != null){
            diff_millseconds= time1.getTime() - this.previous_timestamp.getTime();
        }
        const diff_display = `~${String(diff_millseconds).padStart(6,"0")}ms`
        let time_text = `${time}`.replace("T", " ").substring(0, 19);
        const timestamp = `[${time_text} ${diff_display}] `
        this.previous_timestamp = time1
        return timestamp
    }
}
