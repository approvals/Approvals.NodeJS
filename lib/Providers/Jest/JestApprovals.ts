const StringWriter = require("../../StringWriter");
const approvals = require("../../Approvals");
import { Options } from "../../Core/Options";
import { getJestNamer } from "./JestNamer";


export function verify(sut: any, options?: Options): void {
    const config = approvals.getConfig();
    config.reporters = ["diffmerge"];
    options = options || new Options()
    const writer = new StringWriter(config,  `${sut}`, options.forFile().getFileExtension());
    approvals.verifyWithControl(getJestNamer(), writer, null, config);
}

export function verifyAsJson(data: any, options?: Options): void {
    const text = JSON.stringify(data, null, "  ");
    options = options || new Options()
    options = options.forFile().withFileExtention(".json")
    verify(text, options);
}
