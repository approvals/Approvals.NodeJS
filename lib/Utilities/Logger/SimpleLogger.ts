import {SingleWrapper, ThreadedWrapper, Wrapper} from "./Wrapper";

import {LoggingInstance} from "./LoggingInstance";
import {StringWrapper} from "./StringWrapper";


export class SimpleLogger {


    static _wrapper: Wrapper<LoggingInstance> = new SingleWrapper(new LoggingInstance())

    static show_queries(show: boolean) {
        this._wrapper.get().show_queries(show);

    }

    static show_markers(show: boolean) {
        this._wrapper.get().show_markers(show);
    }

    static show_events(show: boolean) {
        this._wrapper.get().show_events(show);
    }

    static show_messages(show: boolean) {
        this._wrapper.get().show_messages(show);
    }

    static show_variables(show: boolean) {
        this._wrapper.get().show_variables(show);

    }

    static show_hour_glass(show: boolean) {
        this._wrapper.get().show_hour_glass(show);

    }

    static register_logger(log_method: (text:string) => void): void {
        this._wrapper.get().logger = log_method
    }

    static log_to_string(): StringWrapper {
        this._wrapper = new SingleWrapper(new LoggingInstance());
        return SimpleLogger._wrapper.get().log_to_string()

    }

    static use_markers(code: () => void, parameters: (string | (() => string)) = ""): void {
        this._wrapper.get().use_markers(1, code, parameters);
    }

    static variable(name: string, value: any, show_types: boolean = false) {
        SimpleLogger._wrapper.get().variable(name, value, show_types)
    }

    static hour_glass() {
        this._wrapper.get().hour_glass();

    }

    static show_all(show: boolean) {
        this._wrapper.get().show_all(show);
    }

    static event(text: string) {
        this._wrapper.get().event(text);

    }

    static query(queryText: string) {
        this._wrapper.get().query(queryText);
    }

    static message(messageText: string) {
        this._wrapper.get().message(messageText);

    }

    static warning(exception: string | Error) {
        this._wrapper.get().warning(exception)

    }

    static show_timestamps(show: boolean) {
        this._wrapper.get().show_timestamps(show);
    }
}
