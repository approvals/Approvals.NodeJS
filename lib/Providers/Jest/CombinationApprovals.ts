import {verify} from "./JestApprovals";
import {printCombinations, EMPTY} from  "../../Utilities/Printers"

export function verifyAllCombinations1<T1>(func: (i: T1) => any, params1: T1[]) {
    // @ts-ignore
    verify(printCombinations((t1, t2, t3, t4, t5, t6, t7, t8, t9) => func(t1), params1, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations2<T1, T2>(func: (t1: T1, t2: T2) => any, params1: T1[], params2: T2[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: any, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2), params1, params2, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}


export function verifyAllCombinations3<T1, T2, T3>(func: (t1: T1, t2: T2, t3: T3) => any, params1: T1[], params2: T2[], params3: T3[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: any, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3), params1, params2, params3, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations4<T1, T2, T3, T4>(func: (t1: T1, t2: T2, t3: T3, t4: T4) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: any, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4), params1, params2, params3, params4, EMPTY, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations5<T1, T2, T3, T4, T5>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: any, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5), params1, params2, params3, params4, params5, EMPTY, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations6<T1, T2, T3, T4, T5, T6>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: any, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6), params1, params2, params3, params4, params5, params6, EMPTY, EMPTY, EMPTY));
}

export function verifyAllCombinations7<T1, T2, T3, T4, T5, T6,T7>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: any, t9: any) => func(t1, t2, t3, t4, t5, t6,t7), params1, params2, params3, params4, params5, params6, params7, EMPTY, EMPTY));
}
export function verifyAllCombinations8<T1, T2, T3, T4, T5, T6,T7,T8>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7, t8:T8) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[], params8: T8[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: T8, t9: any) => func(t1, t2, t3, t4, t5, t6,t7, t8), params1, params2, params3, params4, params5, params6, params7, params8, EMPTY));
}
export function verifyAllCombinations9<T1, T2, T3, T4, T5, T6,T7,T8,T9>(func: (t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7:T7, t8:T8, t9:T9) => any, params1: T1[], params2: T2[], params3: T3[], params4: T4[], params5: T5[], params6: T6[], params7: T7[], params8: T8[], params9: T9[]) {
    // @ts-ignore
    verify(printCombinations((t1: T1, t2: T2, t3: T3, t4: T4, t5: T5, t6: T6, t7: T7, t8: T8, t9: T9) => func(t1, t2, t3, t4, t5, t6,t7, t8,t9), params1, params2, params3, params4, params5, params6, params7, params8, params9));
}
