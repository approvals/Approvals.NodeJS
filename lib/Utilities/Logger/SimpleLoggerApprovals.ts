import {Options} from "../../Core/Options";
import {SimpleLogger} from "./SimpleLogger";
import {verify} from "../../Providers/Jest/JestApprovals";

export function verifySimpleLogger(testName: string, code: () => void, options?: Options) {
    test(testName, () => {
        const output = SimpleLogger.log_to_string()
        code();
        verify(output, options);
    });
}
