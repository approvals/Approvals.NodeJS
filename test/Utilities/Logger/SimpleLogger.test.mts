import { describe, test } from "@jest/globals";
import { SimpleLogger } from "../../../lib/Utilities/Logger/SimpleLogger.js";
import { verify } from "../../../lib/Providers/Jest/JestApprovals.js";
import { verifySimpleLogger } from "../../../lib/Utilities/Logger/SimpleLoggerApprovals.js";

function logVariables() {
  SimpleLogger.useMarkers(() => {
    const names = ["Jacqueline", "Llewellyn"];
    SimpleLogger.variable("names", names, true);
    SimpleLogger.variable("names", names, false);
  });
}

function methodWithReturnValue(name: string, age: number): string {
  return SimpleLogger.useMarkers(
    () => {
      return `${name} is ${age} years old today.`;
    },
    `name: ${name}, age: ${age}`,
    true,
  );
}

describe("SimpleLogger", () => {
  test("variable with list", () => {
    const output = SimpleLogger.logToString();
    logVariables();
    verify(output);
  });

  verifySimpleLogger("verify", () => {
    logVariables();
  });

  function log_from_inner_method() {
    SimpleLogger.useMarkers(() => {
      const name = "Example";
      SimpleLogger.variable("name", name);
      for (let i = 0; i < 142; i++) {
        SimpleLogger.hourglass();
      }
    });
  }

  verifySimpleLogger("test_standard_logger", () => {
    log_from_inner_method();
  });

  // begin-snippet: verify_simple_logger_example
  verifySimpleLogger("variable", () => {
    SimpleLogger.variable("dalmatians", 101, true);
    SimpleLogger.variable("dalmatians", 101, false);
  });
  // end-snippet

  // begin-snippet: verify_simple_logger_long_example
  test("test_variable_explict", () => {
    const output = SimpleLogger.logToString();
    SimpleLogger.variable("dalmatians", 101, true);
    SimpleLogger.variable("dalmatians", 101, false);
    verify(output);
  });

  // end-snippet

  function verify_toggle(toggle_name: string, toggle: (a: any) => any) {
    SimpleLogger.showAll(true);
    SimpleLogger.event(`Toggle Off ${toggle_name}`);
    toggle(false);
    log_everything();
  }

  function log_everything(): void {
    SimpleLogger.useMarkers(() => {
      SimpleLogger.query("Select * from people");
      SimpleLogger.variable("Nonsense", "foo");
      SimpleLogger.event("Testing");
      SimpleLogger.message("Something random");
      for (let i = 0; i < 13; i++) {
        SimpleLogger.hourglass();
      }
      try {
        throw new Error("Bad stuff happened here");
      } catch (e: any) {
        SimpleLogger.warning(e);
      }
    });
  }

  verifySimpleLogger("switching", () => {
    verify_toggle("None", (_a) => SimpleLogger.showAll(true));
    verify_toggle("All", (a) => SimpleLogger.showAll(a));
    verify_toggle("Query", (a) => SimpleLogger.showQueries(a));
    verify_toggle("Message", (a) => SimpleLogger.showMessages(a));
    verify_toggle("Variable", (a) => SimpleLogger.showVariables(a));
    verify_toggle("Hour Glass", (a) => SimpleLogger.showHourglass(a));
    verify_toggle("Markers", (a) => SimpleLogger.showMarkers(a));
    verify_toggle("Events", (a) => SimpleLogger.showEvents(a));
  });

  // begin-snippet: method_with_inputs_and_outputs
  function method_with_inputs_and_outputs(
    number: number,
    announcement: string,
  ) {
    SimpleLogger.useMarkers(
      () => {
        // end-snippet
        for (number = number; 0 <= number; number--) {
          process.stdout.write(`${number} \n`);
        }
        process.stdout.write(`${announcement}\n`);
      },
      () => `number = ${number}, announcement = ${announcement}`,
    );
  }

  verifySimpleLogger("test_markers_with_signature_in_and_out", () => {
    method_with_inputs_and_outputs(10, "Blast off");
  });

  // begin-snippet: method_with_inputs
  function method_with_inputs(number: number, name: string) {
    SimpleLogger.useMarkers(() => {
      process.stdout.write(`${number}) ${name}`);
    }, `number = ${number}, name = ${name}`);
  }

  // end-snippet

  verifySimpleLogger("test_markers_with_signature", () => {
    method_with_inputs(1, "Susan");
  });

  verifySimpleLogger("test_timestamps", () => {
    let count = -1;

    function create_applesauce_timer(): Date {
      const dates = [
        new Date(0),
        new Date(500),
        new Date(2000),
        new Date(1050000),
        new Date(1052000),
      ];
      count++;
      return dates[count] ?? new Date();
    }

    SimpleLogger._wrapper.get().timer = () => create_applesauce_timer();
    SimpleLogger.showTimestamps(true);
    SimpleLogger.event("1");
    SimpleLogger.event("2");
    SimpleLogger.event("3");
    SimpleLogger.event("4");
    SimpleLogger.warning(new Error("Oh no you didn't!"));
  });

  verifySimpleLogger("test_warnings", () => {
    SimpleLogger._wrapper.get().logStackTraces = true;
    try {
      throw new Error("EVERYTHING IS exceptionally AWFUL!!!!!!");
    } catch (e: any) {
      SimpleLogger.warning(e);
    }
  });

  verifySimpleLogger("Markers with return values", () => {
    methodWithReturnValue("Llewellyn", 29);
  });
});
