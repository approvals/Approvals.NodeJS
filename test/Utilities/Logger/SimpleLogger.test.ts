import {describe, test} from "@jest/globals";
import {SimpleLogger} from "../../../lib/Utilities/Logger/SimpleLogger";
import {verify} from "../../../lib/Providers/Jest/JestApprovals";
import {verifySimpleLogger} from "../../../lib/Utilities/Logger/SimpleLoggerApprovals";

function logVariables() {
    SimpleLogger.useMarkers(() => {
        const names = ["Jacqueline", "Llewellyn"]
        SimpleLogger.variable("names", names, true)
        SimpleLogger.variable("names", names, false)
    });
}


function methodWithReturnValue(name:string, age: number): string {
    return SimpleLogger.useMarkers(() => {
        return `${name} is ${age} years old today.`
    }, `name: ${name}, age: ${age}`, true);

}

describe("SimpleLogger", () => {
    test("variable with list", () => {
        const output = SimpleLogger.logToString()
        logVariables();
        verify(output)
    });

    verifySimpleLogger("verify", () => {
        logVariables();
    });

    function log_from_inner_method() {
        SimpleLogger.useMarkers(() => {

            const name = "Example"
            SimpleLogger.variable("name", name)
            for (let i = 0; i < 142; i++) {
                SimpleLogger.hourglass()
            }
        });

    }

    verifySimpleLogger("test_standard_logger", () => {
        log_from_inner_method()
    });

    // begin-snippet: verify_simple_logger_example
    verifySimpleLogger("variable", () => {
        SimpleLogger.variable("dalmatians", 101, true)
        SimpleLogger.variable("dalmatians", 101, false)
    });
    // end-snippet

    // begin-snippet: verify_simple_logger_long_example
    test("test_variable_explict", () => {
        const output = SimpleLogger.logToString()
        SimpleLogger.variable("dalmatians", 101, true)
        SimpleLogger.variable("dalmatians", 101, false)
        verify(output)
    });

    // end-snippet


    function verify_toggle(toggle_name: string, toggle: (a: any) => any) {
        SimpleLogger.showAll(true)
        SimpleLogger.event(`Toggle Off ${toggle_name}`)
        toggle(false)
        log_everything()
    }

    function log_everything(): void {
        SimpleLogger.useMarkers(() => {

            SimpleLogger.query("Select * from people")
            SimpleLogger.variable("Nonsense", "foo")
            SimpleLogger.event("Testing")
            SimpleLogger.message("Something random")
            for (let i = 0; i < 13; i++) {
                SimpleLogger.hourglass()

            }
            try {
                throw new Error("Bad stuff happened here")
            } catch (e: any) {
                SimpleLogger.warning(e)
            }

        });
    }

    verifySimpleLogger("switching", () => {
        verify_toggle("None", a => SimpleLogger.showAll(true));
        verify_toggle("All", a => SimpleLogger.showAll(a));
        verify_toggle("Query", a => SimpleLogger.showQueries(a));
        verify_toggle("Message", a => SimpleLogger.showMessages(a));
        verify_toggle("Variable", a => SimpleLogger.showVariables(a));
        verify_toggle("Hour Glass", a => SimpleLogger.showHourglass(a));
        verify_toggle("Markers", a => SimpleLogger.showMarkers(a));
        verify_toggle("Events", a => SimpleLogger.showEvents(a));

    });

    // begin-snippet: method_with_inputs_and_outputs
    function method_with_inputs_and_outputs(number: number, announcement:string) {
        SimpleLogger.useMarkers(() => {

                // end-snippet
                for (number = number; 0 <= number; number--) {
                    process.stdout.write(`${number} \n`)
                }
                process.stdout.write(`${announcement}\n`)
            },
            () => `number = ${number}, announcement = ${announcement}`);
    }

    verifySimpleLogger("test_markers_with_signature_in_and_out", () => {
        method_with_inputs_and_outputs(10, "Blast off");
    })

    // begin-snippet: method_with_inputs
    function method_with_inputs(number: number, name:string) {
        SimpleLogger.useMarkers(() => {
            process.stdout.write(`${number}) ${name}`)
        }, `number = ${number}, name = ${name}`);
    }

    // end-snippet

    verifySimpleLogger("test_markers_with_signature", () => {
        method_with_inputs(1, "Susan");
    })

    verifySimpleLogger("test_timestamps", () => {
        let count = -1;

        function create_applesauce_timer(): Date {
            const dates = [
                new Date(0),
                new Date(500),
                new Date(2000),
                new Date(1050000),
                new Date(1052000),
            ]
            count++;
            return dates[count] ?? new Date();
        }

        SimpleLogger._wrapper.get().timer = () => create_applesauce_timer();
        SimpleLogger.showTimestamps(true)
        SimpleLogger.event("1")
        SimpleLogger.event("2")
        SimpleLogger.event("3")
        SimpleLogger.event("4")
        SimpleLogger.warning(new Error("Oh no you didn't!"))
    });

    verifySimpleLogger("test_warnings",
        () => {


            function scrubber(text: string): string {
                return text.replace("", "test_simple_logger.py")
            }

            SimpleLogger._wrapper.get().logStackTraces = true;
            const text = "EVERYTHING IS AWFUL!!!!!!"
            let exception: any;
            try {
                throw new Error("EVERYTHING IS exceptionally AWFUL!!!!!!");
            } catch (e) {
                exception = e;
            }
            //SimpleLogger.warning(text)
            SimpleLogger.warning(exception)
        });

    verifySimpleLogger("Markers with return values", () => {
        methodWithReturnValue("Llewellyn", 29);
    })
});
