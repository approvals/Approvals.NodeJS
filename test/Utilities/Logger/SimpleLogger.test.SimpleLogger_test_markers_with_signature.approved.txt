=> method_with_inputs(number = 1, name = Susan)
<= method_with_inputs()
